### PR TITLE
Add safety module

### PR DESCRIPTION
**Pull request by patil-suraj**
This PR adds `StableDiffusionSafetyChecker` to filter out NSFW content in `StableDiffusionPipeline`.

The `StableDiffusionSafetyChecker` contains the clip vision model, vision projection. The nsfw concept embeddings are pre-computed and part of the state_dict. The module takes images and replaces nsfew image if detected with a black image.

It's added as a new required attribute in `StableDiffusionPipeline` so it'll always be loaded.

### CODE REVIEW

File: src/diffusers/pipeline_utils.py [hunk 1]
```diff
@@ -65,7 +66,9 @@ def register_modules(self, **kwargs):
             # check if the module is a pipeline module
             pipeline_file = module.__module__.split(".")[-1]
             pipeline_dir = module.__module__.split(".")[-2]
-            is_pipeline_module = pipeline_file == "pipeline_" + pipeline_dir and hasattr(pipelines, pipeline_dir)
+            # is_pipeline_module = pipeline_file == "pipeline_" + pipeline_dir and hasattr(pipelines, pipeline_dir)
+            path = module.__module__.split(".")
+            is_pipeline_module = pipeline_dir in path and hasattr(pipelines, pipeline_dir)
```
Comment (patil-suraj) [inline src/diffusers/pipeline_utils.py:71]: More general logic to detect if a module comes from `pipeline` module. For now this is only needed for `LDMBert` model and the safety checker.

This should probably be in a separate PR.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 2]
```diff
@@ -136,7 +148,12 @@ def __call__(
 
         image = (image / 2 + 0.5).clamp(0, 1)
         image = image.cpu().permute(0, 2, 3, 1).numpy()
+
+        # run safety checker
+        safety_cheker_input = self.feature_extractor(self.numpy_to_pil(image), return_tensors="pt").to(torch_device)
+        image, has_nsfw_concept = self.safety_checker(images=image, clip_input=safety_cheker_input.pixel_values)
+
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:155]: `safety_checker` will replace nsfw images (if detected) in `image` with a black image.

**Review by patrickvonplaten — APPROVED** [review]
Great! Looks very nice to me!

**Comment by pcuenca** [comment]
LGTM! I'll test it when I'm added to the `fusing` org.

**Comment by patrickvonplaten** [comment]
> LGTM! I'll test it when I'm added to the `fusing` org.

Added you :-)

**Comment by pcuenca** [comment]
Looks great! Just a couple of quick comments / questions:

- It gives false positives for "A painting of a squirrel eating a burger". I tried a loop with random seeds (set using `torch.manual_seed` before invoking the pipeline) and it consistently said that nsfw content was detected. Perhaps there's some concept in the "be careful" list that's driving this result?
- [Not important for release] How does it impact GPU memory requirements? It has been broadly announced that it runs on GPUs with at least 10 GB of memory -> I'll test in a 11 GB card.
- [Not important] Would it be possible to exclude the safety checker module from the list of modules, so printing the pipeline doesn't reveal it? Probably more trouble than it's worth.

**Comment by patrickvonplaten** [comment]
Regarding `false positives` that's not great indeed @rromb @ablattmann @pesser have you noticed this as well - we should definitely add to the error message that if it's a false positive then the prompt should be re-run with a different seed

**Review by anton-l — APPROVED** [review]
LGTM! Our tests are all based on v1-1, but I've checked manually with the safety checkpoint.
 
One nit: maybe let's add a `logger.warning()` about detected nsfw still, so that it's visible in a terminal/notebooks too.

"The "rerun with a different seed" message can be included in the warning as well, and added to the gradio demo too.

**Comment by BIGJUN777** [comment]
Hi, there. Thank you for releasing the model.
When I tested your model on some images, it turn out that the model tended to output wrong results even though the input images were normal. There are some sample images below. Do you have any idea about this? Thanks.

**Comment by patrickvonplaten** [comment]
Hey @BIGJUN777,

Could you please open a new issue for this?

**Comment by patrickvonplaten** [comment]
It would be nice to have a reproducible code snippet as well
